### DISCUSSION TITLE
error: building aws-c-cal:x64-windows-static-md failed with: BUILD_FAILED

### DISCUSSION THREAD

**Discussion by VikingExplorer** [microsoft/vcpkg#37735] [Q&A] [answered]
I haven't used vcpkg in a year or two. Does anyone have an idea what's going wrong?

I'm using VS2022 17.9.4, v143

C:\Dev\Couloir\3pLibs\vcpkg>.\vcpkg install aws-sdk-cpp[apigateway,apigatewayv2,appsync,autoscaling,autoscaling-plans,budgets,core,dynamodb,ec2,email,glacier,iam,imagebuilder,lambda,lex,lex-models,location,logs,managedblockchain,pricing,rds,rds-data,s3,secretsmanager,sms,sns,sqs,translate,wellarchitected,xray]:x64-windows-static-md --recurse

**warning: vcpkg appears to be in a Visual Studio prompt targeting x86 but installing for x64-windows-static-md. Consider using --triplet x86-windows or --triplet x86-uwp.**

CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: C:/Dev/Couloir/3pLibs/vcpkg/downloads/tools/cmake-3.29.0-windows/cmake-3.29.0-windows-i386/bin/cmake.exe --build . --config Debug --target install -- -v -j17
    Working Directory: C:/Dev/Couloir/3pLibs/vcpkg/buildtrees/aws-c-cal/x64-windows-static-md-dbg
    See logs for more information:
      C:\Dev\Couloir\3pLibs\vcpkg\buildtrees\aws-c-cal\install-x64-windows-static-md-dbg-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:102 (vcpkg_execute_build_process)
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:41 (vcpkg_cmake_build)
  ports/aws-c-cal/portfile.cmake:17 (vcpkg_cmake_install)
  scripts/ports.cmake:175 (include)

error: building aws-c-cal:x64-windows-static-md failed with: BUILD_FAILED
Elapsed time to handle aws-c-cal:x64-windows-static-md: 4.7 s
Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
  https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+aws-c-cal
You can submit a new issue at:
  https://github.com/microsoft/vcpkg/issues/new?title=[aws-c-cal]+Build+error+on+x64-windows-static-md&body=Copy+issue+body+from+C%3A%2FDev%2FCouloir%2F3pLibs%2Fvcpkg%2Finstalled%2Fvcpkg%2Fissue_body.md

**Comment by VikingExplorer** [comment]
**install-x64-windows-static-md-dbg-out.log**

C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1439~1.335\bin\Hostx64\x64\cl.exe   -DDEBUG_BUILD -IC:\Dev\Couloir\3pLibs\vcpkg\buildtrees\aws-c-cal\src\v0.6.10-73dc1009ad.clean\include -external:IC:\Dev\Couloir\3pLibs\vcpkg\installed\x64-windows-static-md\include -external:W0 /nologo /DWIN32 /D_WINDOWS  /utf-8 /MP  /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1  /W4 /WX /MP /volatile:iso /showIncludes /FoCMakeFiles\aws-c-cal.dir\source\cal.c.obj /FdCMakeFiles\aws-c-cal.dir\aws-c-cal.pdb /FS -c C:\Dev\Couloir\3pLibs\vcpkg\buildtrees\aws-c-cal\src\v0.6.10-73dc1009ad.clean\source\cal.c
C:\Dev\Couloir\3pLibs\vcpkg\buildtrees\aws-c-cal\src\v0.6.10-73dc1009ad.clean\include\aws/cal/cal.h(15): error C2220: the following warning is treated as an error
C:\Dev\Couloir\3pLibs\vcpkg\buildtrees\aws-c-cal\src\v0.6.10-73dc1009ad.clean\include\aws/cal/cal.h(15): warning C4431: missing type specifier - int assumed. Note: C no longer supports default-int
C:\Dev\Couloir\3pLibs\vcpkg\buildtrees\aws-c-cal\src\v0.6.10-73dc1009ad.clean\include\aws/cal/cal.h(15): error C2054: expected '(' to follow 'AWS_PUSH_SANE_WARNING_LEVEL'
C:\Dev\Couloir\3pLibs\vcpkg\buildtrees\aws-c-cal\src\v0.6.10-73dc1009ad.clean\include\aws/cal/cal.h(20): error C2296: '<<': not valid as left operand has type 'char [3]'
C:\Dev\Couloir\3pLibs\vcpkg\buildtrees\aws-c-cal\src\v0.6.10-73dc1009ad.clean\include\aws/cal/cal.h(20): error C2297: '<<': not valid as right operand has type 'char [3]'
C:\Dev\Couloir\3pLibs\vcpkg\buildtrees\aws-c-cal\src\v0.6.10-73dc1009ad.clean\include\aws/cal/cal.h(20): error C2056: illegal expression

**Comment by VikingExplorer** [comment, marked as answer]
**Lesson is that packages with Features can't be upgraded. Also, make sure to remove all related packages before attempting to install again.**
 
Won't work: PS C:\Dev\Couloir\3pLibs\vcpkg> ./vcpkg upgrade aws-sdk-cpp[apigateway,apigatewayv2,appsync,autoscaling,autoscaling-plans,budgets,core,dynamodb,ec2,email,glacier,iam,imagebuilder,lambda,lex,lex-models,location,logs,managedblockchain,pricing,rds,rds-data,s3,secretsmanager,sms,sns,sqs,translate,wellarchitected,xray]:x64-windows-static-md
**error: List of features is not allowed in this context**

**My updated procedure:**
UPDATE
1. Open Windows Powershell  
2. cd C:\Dev\Couloir\3pLibs\vcpkg
3. ./bootstrap-vcpkg.bat
4. git pull
5. ./vcpkg update   (to see what is out of date)
6. install or upgrade from PACKAGE LIST  (or all at once: ./vcpkg upgrade --no-dry-run)
---------------------
FIRST TIME
1. install 64 bit git (https://git-scm.com/download/win)
2. Open Windows Powershell
3. cd C:\Dev\Couloir\3pLibs
4. git clone https://github.com/Microsoft/vcpkg

**Reply by dg0yt** [reply]
> **Lesson is that packages with Features can't be upgraded. Also, make sure to remove all related packages before attempting to install again.**

Please don't post wrong "lessons", even if they seem to reflect your experience.

**Reply by VikingExplorer** [reply]
Kai, I don't think you know what is right or wrong.  You've been ZERO help with this issue.  You spew incorrect information, like "vcpkg update [--no-dry-run]"

**Reply by VikingExplorer** [reply]
The problem was that doing a remove on aws-sdk-cpp didn't remove these 3 aws packages:

aws-c-common:x64-windows-static-md 0.4.56 -> 0.9.14
aws-c-event-stream:x64-windows-static-md 0.1.6 -> 0.4.2
aws-checksums:x64-windows-static-md 0.1.9 -> 0.1.18

I had to remove these separately.

**Reply by dg0yt** [reply]
Yes, I should have written `vcpkg upgrade [no-dry-run]`. I will edit my post.

The key point was written in your post:
**error: List of features is not allowed in this context**

You could have solved you original problem with `vcpkg upgrade --no-dry-run` (i.e. update all packages), or even with listing packages. But you cannot add features.

**Reply by VikingExplorer** [reply]
But I absolutely did NOT want to upgrade all my packages.   So, instead of my Lesson being wrong, it was COMPLETELY CORRECT.  FOUR days went by & I had to figure it out myself b/c you were NO help at all.

**Reply by dg0yt** [reply]
I can't help you with the quality and quantity of the Microsoft support. 
IMO some aspects of your issue are special and cannot be generalized. And this doesn't mean they are not reasonable. But it seems we are unable to discuss this at the moment. I will update my post again.

**Comment by dg0yt** [comment]
This answer was edited for corrections and for coverage. It  reflects my understanding of the state of vcpkg, as a community member. It doesn't indicate if I like it or not.

### The `upgrade` command

`vcpkg upgrade ... [--no-dry-run]` updates all packages for which build recipes exist. It is possible to list particular ports, but the command doesn't accept feature specifications.
https://learn.microsoft.com/en-us/vcpkg/commands/upgrade

### Package version control exists only in manifest mode

There is effectively no support for per-port version selection in classic mode. The operation of `vcpkg upgrade` is consistent with this pattern. Strategies to simulate that with manual git operations or sophisticated upgrade commands usually do not offer a convenient and satisfying result or user experience.

Manifest mode has support for selecting version baselines, for pinning versions and for expression minimum versions. Manifest mode doesn't need `vcpkg upgrade'.

Vcpkg CI only tests complete states of the port registry. Regardless of mode, combining ports from an extended range of time results in untested configurations. It may not always be obvious how to fix the resulting problems.

### SDK components may be independent

In classic mode, there is no tracking of whether packages were installed "automatically" (to satisfy dependencies) or on explicit user request. Removing an SDK-like package (`aws-sdk-cpp`) will not remove its dependencies, even if they might be understood as "components" of the SDK. 

In manifest mode, only necessary packages are installed. The manifest *is* the authoritative list of explicitly requested packages. Removing an SDK port from a manfest will also remove its transitive dependencies if they are no longer part of another dependency chain.

**Comment by VikingExplorer** [comment]
Kai, that doesn't seem to be correct.  I don't want people reading this &
thinking that the update command works like that.

https://devblogs.microsoft.com/cppblog/vcpkg-introducing-the-upgrade-command/
https://learn.microsoft.com/en-us/vcpkg/commands/update

If MS had an ounce of sense, they would rename _update_ to _compare_
